### DISCUSSION TITLE
Will the project be updated with SQLAlchemy version 2.0?

### DISCUSSION THREAD

**Discussion by cfraboulet** [fastapi/sqlmodel#547] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
[tool.poetry]
name = "sqlmodel"
version = "0"
description = "SQLModel, SQL databases in Python, designed for simplicity, compatibility, and robustness."
authors = ["Sebastián Ramírez <tiangolo@gmail.com>"]
readme = "README.md"
homepage = "https://github.com/tiangolo/sqlmodel"
documentation = "https://sqlmodel.tiangolo.com"
repository = "https://github.com/tiangolo/sqlmodel"
license = "MIT"
exclude = ["sqlmodel/sql/expression.py.jinja2"]
classifiers = [
    "Development Status :: 4 - Beta",
    "Framework :: AsyncIO",
    "Intended Audience :: Developers",
    "Intended Audience :: Science/Research",
    "Intended Audience :: System Administrators",
    "License :: OSI Approved :: MIT License",
    "Programming Language :: Python :: 3 :: Only",
    "Programming Language :: Python :: 3.6",
    "Programming Language :: Python :: 3.7",
    "Programming Language :: Python :: 3.8",
    "Programming Language :: Python :: 3.9",
    "Topic :: Database",
    "Topic :: Database :: Database Engines/Servers",
    "Topic :: Internet",
    "Topic :: Internet :: WWW/HTTP :: HTTP Servers",
    "Topic :: Internet :: WWW/HTTP",
    "Typing :: Typed",
]

[tool.poetry.dependencies]
python = "^3.6.1"
SQLAlchemy = ">=1.4.17,<=1.4.41"
pydantic = "^1.8.2"
sqlalchemy2-stubs = {version = "*", allow-prereleases = true}
```

### Description

Using FastAPI and SQLModel in my projects, I was wondering if the SQLAlchemy package used in SQLModel was going to be updated to the latest version 2.0.0 which was released on 26 January 2023 in order to use the latest features? 

Thank you very much for your feedback.

### Operating System

Linux

### Operating System Details

Fedora Linux 36

### SQLModel Version

0.0.8

### Python Version

3.9.14

**Comment by tiangolo** [comment, marked as answer]
Yes, it will. ~But there are other things I have to update do first.~

SQLModel 0.0.14 now has support for Pydantic v2, and it also keeps support for Pydantic v1. 🎉

**Reply by cfraboulet** [reply]
Thanks for your answer :) Is it possible to give a hand?

**Reply by arogozhnikov** [reply]
SQLAlchemy 2.0 has ok typing support with pyright (but not other typing tools), integrates with dataclasses and seems to be able to integrate with pydantic [(ref)](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#integrating-with-alternate-dataclass-providers-such-as-pydantic). 

I think those are main points sqlmodel added to sqlalchemy. What would be main features of sqlmodel compared to sqlalchemy2.0?

**Reply by JonathanPlasse** [reply]
Is there a feature like `table=False`?

**Reply by arogozhnikov** [reply]
sorry, coudn't find in sqlmodel documentation what table=False means, can you point me?

**Reply by JonathanPlasse** [reply]
You can either have a sqlmodel that is only a pydantic or a sqlalchemy model for this you have to specify `table` is `False` or `True`
```python
class Model(SQLModel, table=True):
    ...
```

https://sqlmodel.tiangolo.com/tutorial/create-db-and-table/#create-the-table-model-class

**Reply by JonathanPlasse** [reply]
`table=False is the default

**Reply by arogozhnikov** [reply]
If I understand this right, than it's the same as deriving from pydantic Base instead of SQLModel, table=False. Unless there are some other benefits that I don't know about

**Reply by JonathanPlasse** [reply]
You can then inherit it with `table=True` to make it a sqlalchemy model.

**Reply by arogozhnikov** [reply]
ah, ok. This seems to do the same job:

```
class BaseTest(MappedAsDataclass):
    int: PKey


class DerTest(BaseTest, Base):
    other_field: Mapped[str]
```

OTOH there are some things that are tricky/impossible to implement in sqlmodel, like inheritance of table models (IMO cool feature of sqlalchemy).

I'm inclined to migrate to plain sqlalchemy2. So... just curious what are upsides of keeping sqlmodel, hence my question above.

**Reply by 8thgencore** [reply]
it looks like the update will take a long time

**Reply by antont** [reply]
@arogozhnikov did you come to a conclusion there?

I like how seamless the fastapi-sqlalchemy integration is with sqlmodel, but is sure cool how SA 2.0 Mapped and dataclasses support similar things.

**Reply by arogozhnikov** [reply]
I've moved to sqlalchemy2. If you're interested in my personal experience (with pros and cons), I've completely described it here: https://github.com/sqlalchemy/sqlalchemy/discussions/9394

**Reply by antont** [reply]
> sqlalchemy/sqlalchemy#9394

Cool, thanks a lot. Are you using FastAPI also?

**Reply by arogozhnikov** [reply]
just a bit. It is not the main interface for me

**Reply by tiangolo** [reply]
SQLModel 0.0.14 now has support for Pydantic v2, and it also keeps support for Pydantic v1. 🎉

**Comment by oliviernguyenquoc** [comment]
Tried to use SQLALchemy 2.0 without SQLModel (without writing both models and pydantic schemas) in Declerative Mapping style, and it turns out that it's not possible for edge cases.

More info: https://github.com/sqlalchemy/sqlalchemy/discussions/10027

So, IMO SQLModel have still a role to play in a SQLAlchemy 2.0 context

**Reply by antont** [reply]
To me SQLModel also looks useful in how straightforward and clear it makes the basics.

Also, it's small, the main.py being just 655 lines, so I'm also not surprised if it remains a useful glue.

I haven't tried that alternative approach myself though, have only taken a look a couple of times.

**Comment by 8thgencore** [comment]
There is no information about migrating to SQLAlchemy 2.0 and Pydantic v2.

**Reply by martinralfreindl** [reply]
Any updates on this? Would love to FastAPI + SQLModel, but our entire code base is in SQLAlchemy 2.0, so switching isn't possible for us at the moment.

**Reply by ahaly-compass** [reply]
Seems that fastapi and SQLModel are supporting Pydantic2 now. 
Is there still a plan to user SQLAlchemy 2.0 in SQLModel?

**Reply by antont** [reply]
> Seems that fastapi and SQLModel are supporting Pydantic2 now. Is there still a plan to user SQLAlchemy 2.0 in SQLModel?

That was done right before Pydantic2, is out since 0.0.12

> ✨ Upgrade SQLAlchemy to 2.0. PR https://github.com/tiangolo/sqlmodel/pull/700 by [@tiangolo](https://github.com/tiangolo) including initial work in PR https://github.com/tiangolo/sqlmodel/pull/563 by [@farahats9](https://github.com/farahats9).

https://github.com/tiangolo/sqlmodel/releases
